### PR TITLE
Add basic complex functions, fix 2-arg `atan`

### PR DESCRIPTION
**Pull request by simeonschaub**

### CODE REVIEW

File: test/rules/base.jl [hunk 1]
```diff
@@ -1,10 +1,31 @@
-function test_scalar(f, f′, xs...)
+_isapprox(x, y; kwargs...) = isapprox(x, y; kwargs...)
```
Comment (ararslan) [inline test/rules/base.jl]: Why is this necessary?

File: test/rules/base.jl [hunk 2]
```diff
@@ -1,10 +1,31 @@
...
+
+const AbstractDifferential = ChainRules.AbstractDifferential
```
Comment (ararslan) [inline test/rules/base.jl]: This is not how you import specific bindings in Julia. Use `using ChainRules: AbstractDifferential` at the top of `test/runtests.jl` if the binding you want is not already exported.

File: test/rules/base.jl [hunk 3]
```diff
@@ -1,10 +1,31 @@
...
+_isapprox(x::Wirtinger, y; kwargs...) = _isapprox(x.primal, y; kwargs...) &&
+    _isapprox(x.conjugate, Zero(); kwargs...)
+_isapprox(x, y::Wirtinger; kwargs...) = _isapprox(x, y.primal; kwargs...) &&
+    _isapprox(Zero(), y.conjugate; kwargs...)
+_isapprox(x::Wirtinger, y::Wirtinger; kwargs...) = _isapprox(x.primal, y.primal; kwargs...) &&
+    _isapprox(x.conjugate, y.conjugate; kwargs...)
+
+const ≈₁ = _isapprox
+
+function test_scalar(f, f′, xs...; broken = false)
```
Comment (ararslan) [inline test/rules/base.jl]: The `broken` keyword argument doesn't appear to actually be used; it should just be removed.

File: test/rules/base.jl (hunk 1 above)
Comment (simeonschaub) [reply test/rules/base.jl]: Mainly because `isapprox` isn't implemented for `Zero()` and `One()`. It might make sense to overload `Base.isapprox` instead, but I don't know if that belongs in this PR.

File: test/rules/base.jl (hunk 2 above)
Comment (simeonschaub) [reply test/rules/base.jl]: Will do. The problem was that having `using ChainRules: AbstractDifferential` in this file, made the test crash on my machine, for some reason, so this seemed like the quickest fix.

File: test/rules/base.jl (hunk 3 above)
Comment (simeonschaub) [reply test/rules/base.jl]: Yes, that's just a remainder. I'll just delete it.

File: test/rules/base.jl (hunk 1 above)
Comment (ararslan) [reply test/rules/base.jl]: If it's relevant to this change then it's fine in this PR. It's better than adding a separate function at any rate.

File: test/rules/base.jl (hunk 3 above)
Comment (ararslan) [reply test/rules/base.jl]: Sounds good, thanks.

File: test/rules/base.jl (hunk 2 above)
Comment (ararslan) [reply test/rules/base.jl]: Hm, that's... bizarre.

**Comment by simeonschaub** [comment]
@ararslan I've implemented your requested changes. Do you agree with the way `isapprox` behaves now?

**Comment by ararslan** [comment]
I think the `isapprox` overloads should stay in the test utilities.

**Comment by simeonschaub** [comment]
> I think the `isapprox` overloads should stay in the test utilities.

Yes, probably good not to overcomplicate the definition of `AbstractDifferential`s.

File: src/rules/base.jl [hunk 4]
```diff
@@ -64,14 +64,26 @@
...
+@scalar_rule(atan(y, x), @setup(u = x^2 + y^2), (x / u, -y / u))
 @scalar_rule(max(x, y), @setup(gt = x > y), (gt, !gt))
 @scalar_rule(min(x, y), @setup(gt = x > y), (!gt, gt))
 @scalar_rule(mod(x, y), @setup((u, nan) = promote(x / y, NaN16)),
              (ifelse(isint, nan, one(u)), ifelse(isint, nan, -floor(u))))
 @scalar_rule(rem(x, y), @setup((u, nan) = promote(x / y, NaN16)),
              (ifelse(isint, nan, one(u)), ifelse(isint, nan, -trunc(u))))
 
+@scalar_rule(angle(x), @setup(u = abs2(x)), Wirtinger(-im//2 * x' / u, im//2 * x / u))
+frule(::typeof(angle), x::Real) = angle(x), Rule(_ -> Zero())
```
Comment (ararslan) [inline src/rules/base.jl]: Can't you just make this
```julia
@scalar_rule(angle(x::Complex), @setup(u = abs2(x)), Wirtinger(-im//2 * x' / u, im//2 * x / u))
@scalar_rule(angle(x::Real), Zero())
```
?

Comment (simeonschaub) [reply src/rules/base.jl]: I can now!

**Comment by simeonschaub** [comment]
Anything still preventing this from getting merged?

**Comment by oxinabox** [comment]
@simeonschaub sorry for not attending to these earlier, if we had have I think some of the problems @shashi has been running into would already be fixed.

I am away for next 2 weeks. But hopefully someone else can attend to these while I am gone.

**Comment by nickrobinson251** [comment]
I think this just needs rebasing to resolve the conflicts?

**Comment by simeonschaub** [comment]
Something went wrong...

**Comment by oxinabox** [comment]
@simeonschaub  are you using `merge`?
You want to be using `rebase`.
So you don't get a bunch of extra commits in your PR from changes in master.

I am not great at explaining it, but googling 
"How to rebase my PR" hopefully has explaination.

**Comment by simeonschaub** [comment]
@oxinabox Yes, that was probably it. I now just force-pushed it as one commit

File: test/rulesets/Base/base.jl [hunk 5]
```diff
@@ -99,6 +99,18 @@
         end
     end
 
+    @testset "Unary complex functions" begin
+        for x in rand.((Int, Float32, Float64, complex.((Float32, Float64))...))
```
Comment (nickrobinson251) [inline test/rulesets/Base/base.jl]:
```suggestion
        for x in rand.((Int32, Int64, Float32, Float64, Complex{Float32}, Complex{Float64}))
```

Comment (nickrobinson251) [reply test/rulesets/Base/base.jl]: also, is it a problem that rand(Int) is going to give an "extreme" number
```julia
julia> rand(Int32)
507053106

julia> rand(Int64)
7317803238138782059
```

File: src/rulesets/Base/base.jl [hunk 6]
```diff
@@ -83,14 +83,20 @@
 
 @scalar_rule(hypot(x, y), (x / Ω, y / Ω))
 @scalar_rule(sincos(x), @setup((sinx, cosx) = Ω), cosx, -sinx)
-@scalar_rule(atan(x, y), @setup(u = x^2 + y^2), (y / u, -x / u))
```
Comment (nickrobinson251) [inline src/rulesets/Base/base.jl]: this is jsut renaming x -> y, y -> x, right?

Comment (simeonschaub) [reply src/rulesets/Base/base.jl]: Yes, that corresponds better to the geometric meaning

File: test/rulesets/Base/base.jl (hunk 5 above)
Comment (simeonschaub) [reply test/rulesets/Base/base.jl]: Probably best just to remove the integer tests, or at least hard code them to more reasonable values

**Comment by simeonschaub** [comment]
It seems like some of the complex tests need larger tolerances

**Comment by simeonschaub** [comment]
@oxinabox Would it be possible to tag a new release for `ChainRulesCore`, because that seems to be what is breaking the Wirtinger tests right now?

**Comment by oxinabox** [comment]
Tagged, waiting for it to merge.
https://github.com/JuliaRegistries/General/pull/3056

**Comment by oxinabox** [comment]
I have restarted the CI now that the never version of ChainRulesCore has been tagged.

**Comment by simeonschaub** [comment]
Should be ready to merge now
